Ticket/429/artifact/creation

danielsf opened the pull request:
This PR addressed ticket #429. It moves the code to generate the data products needed by the cell_labeling_app into main. As such, it is coupled this PR

https://github.com/AllenInstitute/cell_labeling_app/pull/19

(this ophys_etl PR must be merged before the cell_labeling_app PR can be finished)

I know this looks massive, however: all this PR does is cut-and-paste code that was previously in staging/segmentation_dev into main. We were pretty good about reviewing and testing code before merging it into staging/segmentation_dev. The only real new work here is the placement of various utils in the ophys_etl module.

Validation:
- All of the relevant unit tests have been ported over from staging/segmentation_dev. They run.
- I also ran a version of the cell_labeling_app built from the above PR using artifacts based on this branch (and installing this branch of ophys_etl in the app's backend). The app ran as expected.

danielsf commented:
I acknowledge that this PR and #433 will mean that we have two frameworks for creating video files. The tools were developed in parallel without the foresight that this code would ever get merged over from `staging/segmentation_dev`.

I will open a ticket to evaluate how the two frameworks can be integrated.

Given that they are both tested and, in some ways, are both required by parts of our pipeline (scientist inspection of motion correction for #433; the cell_labeling_app for this PR), I'm going to advocate for merging them both first and combining them later.

I am open to the idea that this is irresponsible behavior.

danielsf commented:
#436 is the ticket to create a unified video generation framework for ophys_etl_pipelines

morriscb commented on tests/utils/test_thumbnail_video_utils.py:
```diff
@@ -0,0 +1,679 @@
+
+@pytest.fixture(scope='session')
+def example_video():
+    rng = np.random.RandomState(16412)
```
use `default_rng` here and throughout file.

morriscb commented on tests/utils/test_thumbnail_video_utils.py:
```diff
@@ -0,0 +1,679 @@
+                             suffix='.h5')[1]
+    rng = np.random.RandomState(714432)
+    with h5py.File(fname, 'w') as out_file:
+        data = rng.randint(0, 65536, size=(214, 10, 10)).astype(np.uint16)
```
iinfo

morriscb commented on tests/utils/test_thumbnail_video_utils.py:
```diff
@@ -0,0 +1,679 @@
+                                          chunks=(100, 5, 5),
+                                          dtype=np.uint16)
+        for chunk in dataset.iter_chunks():
+            arr = rng.randint(0, 65536,
```
iinfo

morriscb commented on tests/utils/test_thumbnail_video_utils.py:
```diff
@@ -0,0 +1,679 @@
+                             min_max,
+                             quantiles,
+                             timesteps):
+    """
```
Okay since this one shows up here. I've noticed that a lot of the functions and tests don't have at least a short, one sentence description of what is being tested or created in the numerous convenience functions. Please add them.

morriscb reviewed (changes requested):
There's a lot of code so I'm sure I didn't catch everything. Main themes are:

There's a lot of `del` statements throughout that seem to be attempting to remove larger objects from memory, however, `del` does not actually clean these from memory and only removes the reference. Unless your goal is to just remove the link, it's likely best to just remove these statements.
Using `iinfo` instead of hand writing the max int.
Changing the deprecated `RandomState` to `default_rng`.
Adding short, single sentence docsstrings to the unittests and convenience functions to explain what is being tested/created.

danielsf commented on src/ophys_etl/utils/video_utils.py:
```diff
@@ -91,3 +92,289 @@ def encode_video(video: np.ndarray, output_path: str,
+        colmax = min(dataset.shape[2], origin[1]+frame_shape[1])
+
+        if quantiles is not None:
+            min_max = np.quantile(dataset[()], quantiles)
```
The thing I am trying to avoid is having two copies of the whole dataset in memory at once.

A few lines down, I need to create a uint8 version of the dataset, and populate it from the normalized chunks of the int16 dataset. Reading in the entire dataset once here isn't necessarily a problem (until we start handling those 100 GB movies I keep fretting about). The problem comes when I construct the empty array of np.uint8s. If this dataset is still in memory, then I will have 2 (really 1.5) complete copies of the movie in memory at once, which can be prohibitive, especially once we start manipulated them in `normalize_array`. Ideally, the user will specify `min_max` rather than `quantiles` in this method, but I don't want to hamstring them this way.

it's not I/O efficient to read the dataset in twice, but it is memory efficient (if only just).

danielsf replied:
I'll add a comment explaining my thinking

danielsf commented on tests/utils/test_thumbnail_video_generator.py:
```diff
@@ -0,0 +1,334 @@
+    assert not expected.video_path == thumbnail.video_path
+    compare_hashes(expected.video_path, thumbnail.video_path)
+
+    del thumbnail
```
I remember why these were there.

We were getting failures where CircleCIs scratch space was filling up with the mp4s that these calls to VideoGenerator.get_thumbnail produce. By calling del on the thumbnail once we were done testing it, we made sure that we didn't fill up the scratch space unnecessarily as tests were run.

I will add comments to that effect so future versions of us know what is going on.

danielsf commented on tests/utils/test_video_normalization_utils.py:
```diff
@@ -0,0 +1,332 @@
+                        geometry['frame_shape'],
+                        quantiles=normalization['quantiles'],
+                        min_max=normalization['min_max'])
+        return
```
For the record, pytest will tell you which combo was the problem. Running this
```
import pytest
from itertools import product

@pytest.mark.parametrize('x, y, z', product((1,2,3), (1,2,3), (1,2,3)))
def test_silliness(x, y, z):
    if x==y and y==z:
        raise RuntimeError("oh no")
    pass
```
gives
```
....
____________________________ test_silliness[3-3-3] _____________________________

x = 3, y = 3, z = 3

    @pytest.mark.parametrize('x, y, z', product((1,2,3), (1,2,3), (1,2,3)))
    def test_silliness(x, y, z):
        if x==y and y==z:
>           raise RuntimeError("oh no")
E           RuntimeError: oh no

test_pytest_msg.py:7: RuntimeError
=========================== short test summary info ============================
FAILED test_pytest_msg.py::test_silliness[1-1-1] - RuntimeError: oh no
FAILED test_pytest_msg.py::test_silliness[2-2-2] - RuntimeError: oh no
FAILED test_pytest_msg.py::test_silliness[3-3-3] - RuntimeError: oh no
========================= 3 failed, 24 passed in 0.11s =========================
```

Granted, it's not clear that's what I was thinking here. I just wanted to let you know that pytest does give you a fairly detailed list of which parametrized tests failed.

morriscb replied:
Either way. I think the internals of the test can be a bit complex, especially testing for the `raises` conditions.

danielsf commented on src/ophys_etl/utils/video_utils.py:
```diff
@@ -91,3 +92,289 @@ def encode_video(video: np.ndarray, output_path: str,
+            data = read_data[:,
+                             origin[0]:origin[0]+frame_shape[0],
+                             origin[1]:origin[1]+frame_shape[1]]
+            del read_data
```
I fixed this one, but I'm going to leave most of them. As per our discussion on Slack, my experience has been that the python garbage collector is good about picking things up once their last reference is dropped.

danielsf commented on tests/modules/roi_cell_classifier/test_labeler_artifact_module.py:
```diff
@@ -0,0 +1,230 @@
+        raw_video = np.round(255.0*raw_video/delta).astype(np.uint8)
+        np.testing.assert_array_equal(raw_video, scaled_video)
+        del raw_video
+        del scaled_video
```
See my paranoia below about filling up scratch on the machines CircleCI is giving us.

danielsf commented:
@morriscb I've made most of the changes you requested. The exceptions, which we discussed on Slack are

- I'm leaving in my invocations of 255; I think `(255, 0, 0)` is more readily identifiable as "red" than `(np.iinfo(np.uint8).max, 0, 0)`
- I'm leaving in most of the calls to `del`. My experimentation on my laptop and on Linux indicates that the garbage collector is pretty aggressive and does free up memory once the last reference goes out of scope.

Along that vein, I added this commit, which might be contentious

https://github.com/AllenInstitute/ophys_etl_pipelines/pull/435/commits/e23d7c904e1a4466bfcb0a6f55f65fc511d00ae5

`VideoGenerator` creates a temporary directory where it can store any `ThumbnailVideo`s it generates. Ideally, `VideoGenerator` would clean up after itself, but there was no way to guarantee that the `ThumbnailVideo`s went out of scope before `VideoGenerator` did. I fixed the problem by having each `ThumbnailVideo` carry around a reference to the generator that created it. Now, the garbage collector will only come for a `VideoGenerator` after all of its `ThumbnailVideo`s have been collected.

Feel free to push back on this. The one risk I can see is that, if one is using a `VideoGenerator` that carries `video_data` (an array) rather than `video_path` (a path to an HDF5 file) around, the array won't get cleaned up until all of the `ThumbnailVideo`s have been cleaned up. Probably that is an edge case, though, and I don't like the current behavior, with `VideoGenerator` leaving an empty directory behind in its wake.

morriscb commented:
Gave you the approval for merging. Last thing I don't know if I saw was adding docstrings to tests.

danielsf commented:
@morriscb Is this the kind of thing you wanted?

https://github.com/AllenInstitute/ophys_etl_pipelines/pull/435/commits/e340e3155157ec431a04bce93733ec14d0170490
https://github.com/AllenInstitute/ophys_etl_pipelines/pull/435/commits/55e826f0d78fd87267837357a9bc4b866ea9d0a0
